pkg/action/upgrade.go

```diff
@@ -378,43 +381,97 @@ func (u *Upgrade) releasingUpgrade(c chan<- resultMessage, upgradedRelease *rele
-				u.reportToPerformUpgrade(c, upgradedRelease, results.Created, err)
-				return
+	deployedResourcesCalculator := phasemanagers.NewDeployedResourcesCalculator(history, u.StagesSplitter, u.cfg.KubeClient)
+
+	rolloutPhaseManager, err := phasemanagers.
+		NewRolloutPhaseManager(rolloutPhase, deployedResourcesCalculator, upgradedRelease, u.cfg.Releases, u.cfg.KubeClient).
+		AddPreviouslyDeployedResources(toBeAdopted).
+		AddCalculatedPreviouslyDeployedResources()
+	if err != nil {
+		u.cfg.recordRelease(originalRelease)
+		u.reportToPerformUpgrade(c, upgradedRelease, kube.ResourceList{}, fmt.Errorf("error calculating previously deployed resources for rollout phase manager: %w", err))
+		return
+	}
+
+	if err := rolloutPhaseManager.DoStage(
+		func(stgIndex int, stage *stages.Stage, prevDeployedStgResources kube.ResourceList) error {
+			if len(prevDeployedStgResources) == 0 {
+				stage.Result, err = u.cfg.KubeClient.Create(stage.DesiredResources)
+				if err != nil {
+					return err
+				}
+			} else {
+				stage.Result, err = u.cfg.KubeClient.Update(prevDeployedStgResources, stage.DesiredResources, u.Force)
+				if err != nil {
+					return err
+				}
 			}
-		} else {
-			if err := u.cfg.KubeClient.Wait(target, u.Timeout); err != nil {
-				u.cfg.recordRelease(originalRelease)
-				u.reportToPerformUpgrade(c, upgradedRelease, results.Created, err)
-				return
+
+			if u.Recreate {
+				// NOTE: Because this is not critical for a release to succeed, we just
+				// log if an error occurs and continue onward. If we ever introduce log
+				// levels, we should make these error level logs so users are notified
+				// that they'll need to go do the cleanup on their own
+				if err := recreate(u.cfg, stage.Result.Updated); err != nil {
+					u.cfg.Log(err.Error())
+				}
 			}
+
+			if u.Wait {
+				if u.WaitForJobs {
+					if err := u.cfg.KubeClient.WaitWithJobs(stage.DesiredResources, u.Timeout); err != nil {
+						return err
+					}
+				} else {
+					if err := u.cfg.KubeClient.Wait(stage.DesiredResources, u.Timeout); err != nil {
+						return err
+					}
+				}
+			}
+
+			return nil
+		},
+	); err != nil {
+		u.cfg.recordRelease(originalRelease)
+
+		var createdResources kube.ResourceList
+		if lastStgWithResult := rolloutPhaseManager.Phase.SortedStages.LastWithResult(); lastStgWithResult != nil {
+			createdResources = lastStgWithResult.Result.Created
 		}
+
+		u.reportToPerformUpgrade(c, upgradedRelease, createdResources, fmt.Errorf("error processing rollout phase stage: %w", err))
+		return
+	}
+
+	if err := rolloutPhaseManager.DeleteOrphanedResources(); err != nil {
+		u.cfg.Log("failure removing resources no longer present in the release: %w", err)
 	}
 
 	// post-upgrade hooks
 	if !u.DisableHooks {
+		var createdResources kube.ResourceList
+		if lastStgWithResult := rolloutPhaseManager.Phase.SortedStages.LastWithResult(); lastStgWithResult != nil {
+			createdResources = lastStgWithResult.Result.Created
+		}
```

Comment:
Created resources from previous stages are ignored here. There are more places we do it like that, same with the Updated/Deleted resources.